### PR TITLE
Add pagination for transaction resource

### PR DESCRIPTION
**Pull request by p-janik**
Todo:
- [x] update kontist.dev sdk page reagrding changes in this PR

### CODE REVIEW

File: lib/graphql/resultPage.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
+import { PageInfo } from "./schema";
+import { IFetch } from "./interfaces";
+
+export class ResultPage {
+  items: Array<Object>;
+  nextPage?: () => Promise<ResultPage>;
+  previousPage?: () => Promise<ResultPage>;
+
+  constructor(model: IFetch, items: Array<Object>, pageInfo: PageInfo) {
+    this.items = items;
```
Comment (johannespfeiffer) [inline lib/graphql/resultPage.ts]: Any chance we can use generics here? 
```
export class ResultPage<T> {
  items: Array<T>;
  nextPage?: () => Promise<ResultPage<T>>;
  previousPage?: () => Promise<ResultPage<T>>;

  constructor(model: IFetch, items: Array<T>, pageInfo: PageInfo) {
```

then later in `transactions.ts` sth. like

```
export class Transaction extends Model {
   async fetch(args?: FetchOptions): Promise<ResultPage<Transaction>> {
```

File: lib/graphql/transaction.ts [hunk 2]
```diff
@@ -5,50 +5,58 @@ import {
   TransactionsConnectionEdge
 } from "./schema";
 import { Model } from "./model";
+import { FetchOptions } from "./types";
+import { ResultPage } from "./resultPage";
+
+const safeGet = (source: any, path: any, defaultValue: any) =>
+  get(source, path, defaultValue) || defaultValue;
```
Comment (johannespfeiffer) [inline lib/graphql/transaction.ts]: Just out of curiosity: When do we need that?

File: lib/graphql/transaction.ts [hunk 3]
```diff
@@ -5,50 +5,58 @@ import {
...
-                  mandateNumber
-                  paymentMethod
-                  category
-                  userSelectedBookingDate
-                  purpose
-                  bookingType
-                  invoiceNumber
-                  invoicePreviewUrl
-                  invoiceDownloadUrl
-                  documentType
```
Comment (johannespfeiffer) [inline lib/graphql/transaction.ts]: Did we loose them? Should be `documentNumber`, `documentPreviewUrl`, `documentDownloadUrl`, ...

File: lib/graphql/types.ts [hunk 4]
```diff
@@ -0,0 +1,6 @@
+export type FetchOptions = {
+  first?: number | null;
+  last?: number | null;
+  before?: String | null;
+  after?: String | null;
```
Comment (johannespfeiffer) [inline lib/graphql/types.ts]: Why do we allow `number` and `null` as a type? The `?` already allows the type to be `undefined`.

Comment (johannespfeiffer) [inline lib/graphql/types.ts]: I think we can use just `string` instead of `String`.

File: lib/graphql/transaction.ts [hunk 5]
```diff
@@ -5,50 +5,58 @@ import {
...
-  async fetchAll(): Promise<Array<TransactionEntry>> {
-    const query = `{
-      viewer {
+  async fetch(args?: FetchOptions): Promise<ResultPage> {
+    const query = `
+      query fetchTransactions ($first: Int, $last: Int, $after: String, $before: String) {
+        viewer {
           mainAccount {
-            transactions {
+            transactions(first: $first, last:$last, after: $after, before: $before) {
```
Comment (johannespfeiffer) [inline lib/graphql/transaction.ts]: nit: missing space in `last:$last`

File: lib/graphql/model.ts [hunk 6]
```diff
@@ -1,13 +1,16 @@
...
   protected client: GraphQLClient;
 
   constructor(graphqlClient: GraphQLClient) {
     this.client = graphqlClient;
   }
 
-  async fetchAll(): Promise<Array<Object>> {
+  async fetch(args?: FetchOptions): Promise<ResultPage> {
     throw new Error("Must be overridden by subclasses!");
   }
```
Comment (johannespfeiffer) [inline lib/graphql/model.ts]: Can we make the class and this method abstract instead?

File: lib/graphql/transaction.ts (hunk 2 above)
Comment (p-janik) [reply lib/graphql/transaction.ts]: `get` returns `defaultValue` only when property from patch is undefined. Does not work like this in other falsy cases

File: lib/graphql/transaction.ts (hunk 3 above)
Comment (p-janik) [reply lib/graphql/transaction.ts]: true, it was renamed

File: lib/graphql/types.ts (hunk 4 above)
Comment (p-janik) [reply lib/graphql/types.ts]: blindly followed `before` and `after`

**Comment by p-janik** [comment]
Thanks, created https://github.com/kontist/kontist.github.io/pull/12 to update the docs

**Pull request merged**